eliot-service/eliot/health_resource.py

```diff
@@ -0,0 +1,110 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+"""
+Application-health related Falcon resources.
+"""
+
+from collections import OrderedDict
+import json
+
+from dockerflow.version import get_version
+import falcon
+import markus
+
+
+METRICS = markus.get_metrics(__name__)
+
+
+class BrokenResource:
+    """Handle ``/__broken__`` endpoint."""
+
+    def on_get(self, req, resp):
+        """Implement GET HTTP request."""
+        METRICS.incr("broken.count")
+        # This is intentional breakage
+        raise Exception("intentional exception")
+
+
+class VersionResource:
+    """Handle ``/__version__`` endpoint."""
+
+    def __init__(self, basedir):
+        self.basedir = basedir
+
+    def on_get(self, req, resp):
+        """Implement GET HTTP request."""
+        METRICS.incr("version.count")
+        version_info = json.dumps(get_version(self.basedir) or {})
+        resp.content_type = "application/json; charset=utf-8"
+        resp.status = falcon.HTTP_200
+        resp.body = version_info
+
+
+class LBHeartbeatResource:
+    """Handle ``/__lbheartbeat__`` to let the load balancing know application health."""
+
+    def on_get(self, req, resp):
+        """Implement GET HTTP request."""
+        METRICS.incr("lbheartbeat.count")
+        resp.content_type = "application/json; charset=utf-8"
+        resp.status = falcon.HTTP_200
+
+
+class HealthState:
+    """Object representing health of system."""
+
+    def __init__(self):
+        self.errors = []
+        self.statsd = {}
+
+    def add_statsd(self, name, key, value):
+        """Add a key -> value gauge."""
+        if not isinstance(name, str):
+            name = name.__class__.__name__
+        self.statsd["%s.%s" % (name, key)] = value
+
+    def add_error(self, name, msg):
+        """Add an error."""
+        # Use an OrderedDict here so we can maintain key order when converting
+        # to JSON.
+        d = OrderedDict([("name", name), ("msg", msg)])
+        self.errors.append(d)
+
+    def is_healthy(self):
+        """Return whether this represents a healthy state."""
+        return len(self.errors) == 0
+
+    def to_dict(self):
+        """Convert state to a dict."""
+        return OrderedDict([("errors", self.errors), ("info", self.statsd)])
```

Comment:
We don't need this. It was something I copied from Antenna which is a bit more complex.